doc/jpgtn.1
.\" -*- nroff -*-
.TH JPGTN 2 "June 14, 2002" "Version 2.01" "Graphics Software"

.SH NAME
jpgtn \- create thumbnails of JPEG files 

.SH SYNOPSIS
.TP
.B jpgtn [options] file(s)

.SH OPTIONS
.TP
.B \-d <directory>
Specifies a directory to put the output file(s) in. If this option is not 
supplied and the -S option is not in effect, it defaults to the current 
directory.
.TP
.B \-f
Force overwrites of existing files when an output file already exists.
.TP
.B \-h
Print a short usage message and exit.
.TP
.B \-p <prefix>
Specify the filename prefix used for output files.
.TP
.B \-q <quality>
Set JPEG quality setting for output file. The
.B <quality>
parameter is a number between 0 and 100. Effective range is probably only 
between 15 and 90.
.TP
.B \-s <size>
The default behavior of this switch is to specify the size of the longest 
dimension of the output images in pixels. For
instance, a 
.B size
of 75 would create a thumbnail of the original image that would fit inside a 
75 X 75 pixel box. The thumbnail's length/width ratio will be the same as the 
original image's. When used in conjunction with the 
.B \-H
switch or the 
.B \-W
switch, this option specifies the height or width of the final thumbnail. 
If this option is not supplied, or if it is not in the allowable range 
(2-2048) the size defaults to 128. The allowable range can be altered only 
prior to compiling.
.TP
.B \-v
Increase the verbosity level. This switch may be used multiple times. 
.TP
.B \-H
Specify that the thumbnail height should be the size specified with the 
.B \-s
switch (see above.) In this case, the thumbnail width is calculated by 
.B jpgtn
and may be greater or lesser than the size specified by the 
.B \-s 
switch depending on the original image.
.TP
.B \-S
Send output to STDOUT instead of file(s). If this option is supplied, the -p 
and -d options are ignored and jpgtn only processes the first file on the 
command line.
.TP
.B \-W
Specify that the thumbnail width should be the size specified with the 
.B \-s
switch (see above.) In this case, the thumbnail height is calculated by 
.B jpgtn
and may be greater or lesser than the size specified by the 
.B \-s 
switch depending on the original image.
.TP
.B \-V
Print the version and copyright banner and exit.

.SH DESCRIPTION
.B Jpgtn
creates small versions of pictures stored in JPEG format and outputs these 
thumbnails either to STDOUT (for use with pipes, file redirection, or CGI 
scripts) or to the original filename with an optional prefix. You may also
specify a directory which you wish all the output files to be written to.
.PP
Jpgtn writes one output file for each JPEG image file specified on the command
line. In order to be able to associate these output files with the input files
that they are generated from, jpgtn names them based on the input filename. 
In order to avoid accidental overwriting of the input files, jpgtn allows 
you to specify a prefix to prepend to each input filename to construct the 
output filename. Jpgtn also allows you to specify a directory to write the 
output files to. If no directory and no prefix is specified, jpgtn writes the
output files to the current directory using names constructed by prepending the 
default prefix "tn_" to the input filenames. 
.PP
If no prefix is specified, but a directory 
.I is 
specified, jpgtn will write one file for each input file to the directory. 
The resultant files will have the same filenames as the original images they
were generated from.  
.PP
If no directory is specified, but a prefix 
.I is 
specified, then jpgtn writes its output files to the current directory and 
names its output files by prepending the supplied prefix to the filename 
of the original image. 
.PP
If 
.I both
a directory and a prefix are specified, them jpgtn writes its output files to 
the specified directory
.I and
prepends the supplied prefix to the original image's filename to construct
the output file's name.
.PP
The 
.B \-S 
option forces jpgtn to write its output to stdout. If this option is 
supplied, jpgtn ignores the 
.B \-d
and the 
.B \-p
options. This option is useful if you want to take advantage of output 
redirection or if you want to dynamically generate thumbnails for a CGI
application. 
.PP 
The \-s
option gives you some control over the size of the thumbnails that are 
generated. If you do not supply this option, a default size of 128 pixels 
is assumed. The number of pixels in the longest dimension of a thumbnail 
generated by jpgtn will not exceed 'size' pixels. E.g. if you supply this
option with a parameter of 100 while making a thumbnail of a 200x100 pixel
image, the resulting image will be 100x50 pixels. If the original jpeg is 
200x200 pixels, the thumbnail will be 100x100, and so on.
.PP
The
.B -q
option sets the JPEG quality level of the output JPEG file. This
defaults to 70, and although values between 1 and 100 inclusive are
allowed, I would suggest that 15 is probably as low as you can go
without running into readability problems etc., and there isn't much
point going above 90. Also, note that quality level 100 is not
lossless, nor is the quality level a percentage. See the JPEG FAQ for
further enlightenment.

.SH EXAMPLES
.PP
To create thumbnails of all the '.jpg' files in the current directory and 
output them to the current directory with the default prefix "tn_" use:

$ jpgtn *.jpg

To do the same thing as above, but to put the thumbnails in './thumbs' use:

$ jpgtn -d "./thumbs/" -p "tn_" *.jpg

To create a thumbnail of the file image.jpg no larger than 75x75 pixels and 
pipe it through the (hypothetical) program "jpegprog" use:
 
$ jpgtn -S -s 75 image.jpg | jpegprog

To create a thumbnail which has a width of 64 pixels and maintains the 
original's height/width ratio from a file called foo.jpg use:

$ jpgtn -W -s 64 foo.jpg

To create a thumbnail which has a height of 64 pixels and maintains the 
original's height/width ratio from a file called bar.jpg use:

$ jpgtn -H -s 64 bar.jpg

.SH ACKNOWLEDGEMENTS
.PP
Thanks to the Independent JPEG Group for writing their useful and flexible 
JPEG software, without which this program could not use the JPEG/JFIF file 
format.
.PP
Thanks to Willie Daniel who created 
.B gtnpic 
which this program is derived from. Thanks to Russell Marks who created 
.B tnpic 
which was the foundation for 
.B gtnpic 
and therefore the foundation of 
.B jpgtn
as well.
.PP
Thanks to Terry Mackintosh for suggesting that there should be a way to make
.B jpgtn
resize a specified dimension while allowing the other dimension to float.

.SH BUGS
Probably has some... if you find one please inform me at the address below.
.SH AUTHOR
Jeremy Madea <jpgtn@madea.net>
.SH SEE ALSO
.BR " xv "(1), " cjpeg "(1), " djpeg "(1)"


NEWS
New in 2.06: 
* Man page is now installed correctly.

New in 2.05: 
* No changes seen by the end user.

New in 2.04:
* No changes seen by the end user. 

New in 2.03:
* --with-libjpeg=PATH configure option. 

New in 2.02:
* GNU autoconf and automake driven build process
* The directory structure for the package has changed.


src/jpgtn.c
/*
    Copyright (C) 2002 Jeremy Madea <jeremymadea@mindspring.com>

    This file is part of jpgtn.

    Jpgtn is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

    Jpgtn is based on two programs: tnpic (distributed with zgv) and gtnpic.
    The original "gtnpic" program was Copyright (C) 1997      Willie Daniel.
    The original "tnpic"  program was Copyright (C) 1993-1996 Russell Marks.
    Both tnpic and gtnpic were distributed under the GNU Public License.
*/

#if HAVE_CONFIG_H
#    include <config.h>
#endif

#include "common.h"
#include "jpgtn.h"
#include <sys/types.h> //included to open directory
#include <dirent.h>

typedef struct node {
    char *path;
    struct node * next;
} node;

node* head = NULL;
node* cursor = NULL;


/* Private */
static void print_version(void);
static void print_usage(char *);
static int  file_exists(const char *);
static char *output_file_name(const char *,int,const char *,int,const char *);
static void get_files(char *);
static void insert_into_array(char *);



int main(int argc, char **argv)
{
    int f;                                 /* Current file to be indexed  */
    int xsiz = 0;                          /* Total output image xsize    */
    int ysiz = 0;                          /* Total output image ysize    */
    unsigned char *bigimage   = NULL;      /* Pointer to current image    */
    unsigned char *palrgb     = NULL;      /* Pointer to image palette    */
    int x=0;
    int  size = DEFAULTSIZE;               /* Size of thumbnail image.    */
    int  opt_size;
    int  opt_flag_quality   = 0;
    int  opt_flag_prefix    = 0;
    int  opt_flag_directory = 0;
    int  opt_flag_stdout    = 0;
    int  opt_flag_force     = 0;
    int  opt_flag_verbose   = 0;
    int  opt_quality;                      /* JPEG Quality                */
    int  opt_rsz_dim = RSZ_AUTO;           /* Dimension to resize. 0,1,2  */
    char *opt_directory = NULL;            /* Output directory            */
    char *opt_prefix    = NULL;            /* Output filename prefix      */
    char *imgfile       = NULL;            /* Output filename             */

    int grid_mode = 0;
    char *source_directory = NULL;
    int grid_width = 0;
    int grid_height = 0;
    int src_directory = 0;
    int rows_per_image = 0;
    int columns_per_image = 0;



    /* Check command args */
    if (argc < 2) {
        print_usage(argv[0]);
        exit(EXIT_FAILURE);
    }

    while (-1 != (x = getopt(argc,argv,"Vh?SHGWfvs:q:d:p:R:C:J:K:D:"))) {
        switch(x) {
            case 'V':
                print_version();
                exit(EXIT_FAILURE);
                break;
            case 'h':
                print_usage(argv[0]);
                exit(EXIT_FAILURE);
                break;
            case 'q':
                sscanf(optarg,"%d",&opt_quality);
                if(opt_quality < 1 || opt_quality > 100) {
                    opt_quality = 70;
                }
                opt_flag_quality = 1;
                break;
            case 's':
                sscanf(optarg,"%d",&opt_size);
                if ((opt_size < MINSIZE) || (opt_size > MAXSIZE)) {
                    opt_size = DEFAULTSIZE;
                }
                size = opt_size;
                break;
            case 'f':
                opt_flag_force = 1;
                break;
            case 'v':
                opt_flag_verbose++;
                break;
            case 'S':
                opt_flag_stdout = 1;
                break;
            case 'H':
                opt_rsz_dim = RSZ_HEIGHT;
                break;
            case 'w':
                opt_rsz_dim = RSZ_WIDTH;
                break;
            case 'd':
                opt_directory = optarg;
                opt_flag_directory = 1;
                break;
            case 'p':
                opt_prefix = optarg;
                opt_flag_prefix = 1;
                break;
            case 'D':
                source_directory = optarg;
                src_directory = 1;
                break;
            case 'R':
                sscanf(optarg,"%i",&rows_per_image);
                break;
            case 'C':
                sscanf(optarg,"%i",&columns_per_image);
                break;
            case 'J':
                sscanf(optarg,"%i",&grid_height);
                break;
            case 'K':
                sscanf(optarg,"%i",&grid_width);
                break;
            case 'G':
                grid_mode = 1;
                break;

            case '?':
                print_usage(argv[0]);
                exit(EXIT_FAILURE);
                break;
            default:
                fprintf(stderr,"Unknown error. Exiting.\n");
                exit(EXIT_FAILURE);
                break;
            }
        }

    if ((!opt_flag_stdout) && (!opt_flag_directory) && (!opt_flag_prefix)) {
        opt_flag_prefix = 1;
        opt_prefix = "tn_";
    }

    if (!opt_flag_quality) opt_quality = 70;

    if (opt_flag_verbose > 2) {
        printf("quality: %d\n",opt_quality);
    }

    if(grid_mode == 1) {
        //grid mode enabled
        //TODO: check parameters existence

        //get images
        if(src_directory == 1) {

            get_files(source_directory);
            for(node* i = head; i != NULL; i = i->next ){
                printf("%s\n", i->path);
            }
        }



    } else {
        if (argc - optind < 1) {
            print_usage(argv[0]);
            exit(EXIT_FAILURE);
        }
        /* The big loop */
        for (f = optind; f < argc; f++) {

            if (!opt_flag_stdout) {
                imgfile = output_file_name(argv[f],
                                           opt_flag_directory, opt_directory,
                                           opt_flag_prefix, opt_prefix);
                if (imgfile != NULL) {
                    if (!opt_flag_force && file_exists(imgfile)) {
                        fprintf(stderr,"Skipping %s: File exists\n", imgfile);
                        continue;
                    }
                } else {
                    fprintf(stderr, "Could not allocate space for filename!\n");
                    exit(EXIT_FAILURE);
                }
            }

            /* Get a pointer to the big image or NULL on error */
            bigimage=read_JPEG_file(argv[f], &palette);
            if (NULL == bigimage) {
                fprintf(stderr, "read_JPEG_file: Error\n");
                if(outimage) free(outimage);
                exit(EXIT_FAILURE);
            }
            palrgb=palette;

            /* Returns a pointer to the resized image or NULL on error */
            outimage = resizepic(bigimage, palrgb, palrgb+256, palrgb+512,
                                 width, height, size, opt_rsz_dim);
            xsiz = out_wide;
            ysiz = out_high;

            if(outimage == NULL) {
                fprintf(stderr,"outimage = NULL ???\n");
                if(bigimage) free(bigimage);
                exit(EXIT_FAILURE);
            }

            if ((xsiz < MINSIZE) && (ysiz < MINSIZE)) {
                 fprintf(stderr, "thumbnail too small...\n");
            }

            free(palrgb);
            free(bigimage);

            if (! opt_flag_stdout) {
                if (opt_flag_verbose) {
                    printf("%s",imgfile);
                    if (opt_flag_verbose > 1) printf(" %dx%d",xsiz,ysiz);
                    printf("\n");
                }
                write_JPEG_file(imgfile,xsiz,ysiz,opt_quality);
                free(imgfile);
                imgfile = NULL;
                free(outimage);

            } else {
                write_JPEG_file(NULL,xsiz,ysiz,opt_quality);
                exit(EXIT_SUCCESS);
            }
        }
    }




    exit (EXIT_SUCCESS);
}

static int file_exists(const char *path) {
    struct stat buf;

    /* Crude check. If we can stat it successfully, we assume it exists. */
    /* Otherwise we return zero, meaning that this function can't tell. */

    if (0 == stat(path, &buf)) {
        return(1);
    } else {
        return(0);
    }
}

static char *output_file_name(const char *infile,
                              int dflag, const char *dir,
                              int pflag, const char *pre)
{
    int   len       = 0;
    char *infiledup = NULL;
    char *ifn       = NULL;
    char *name      = NULL;

    /* Make a copy of the input file name. */
    infiledup = strdup(infile);
    if (! infiledup) {
        return(NULL);
    }

    /* Ignore path information at the beginning of the filename. */
    ifn = strrchr(infiledup,'/');
    if (!ifn) {
        ifn = infiledup;
    } else {
        ifn++;
    }

    /* Estimate max length we'll need; add two for "./" and one for null. */
    len = strlen(ifn) + 3;
    if (dir) len += strlen(dir);
    if (pre) len += strlen(pre);

    /* Reserve space for it... */
    name = (char *)malloc(len);

    if (name == NULL) {
        return NULL;
    }

    *name = '\0';

    /* Construct the output file name. */
    if (dflag) {
       strcpy(name, dir);
       len = strlen(dir);

        /* Add "/" on dir if needed.*/
        if (dir[len - 1] != '/') {
            strcat(name,"/");
        }
    } else {
        /* strcpy(name, "./"); */
    }

    if (pflag) {
        strcat(name, pre);
        strcat(name, ifn);
    } else {
        strcat(name, ifn);
    }

    free(infiledup);
    return(name);
}

static void print_usage(char *prog)
{
    print_version();
    printf("usage: %s [switches] [inputfiles]\n",prog);
    printf("  -d directory  Set the output directory.\n");
    printf("  -f            Force overwrite of existing output files.\n");
    printf("  -h            Display this help message and exit.\n");
    printf("  -p prefix     Set the output filename prefix.\n");
    printf("  -q quality    Set the output quality.\n");
    printf("  -s size       Set the size of the longest output dimension. ");
    printf(" (See -H and -V)\n");
    printf("  -v            Verbosity level. (May be used multiple times.)\n");
    printf("  -H            Make the -s switch refer to the output height.\n");
    printf("  -S            Process only one file and output to stdout.\n");
    printf("  -W            Make the -s switch refer to the output width.\n");
    printf("  -G            Set the grid mode.\n");
    printf("  -D directory  Set the source directory.\n");
    printf("  -R            Set the number of rows in the grid.\n");
    printf("  -C            Set the number of columns in the grid.\n");
    printf("  -J            Set the height of the grid image.\n");
    printf("  -K            Set the width of the grid image.\n");
    printf("  -V            Display version and exit.\n");
    printf("\nSee jpgtn(1) for more details.\n\n");
}

static void print_version(void)
{
    printf("jpgtn : Version %s\n",VERSION);
    printf("Copyright (c) 2002 Jeremy Madea <jeremymadea@mindspring.com>\n\n");
}

static void get_files(char *folder_name) {
    DIR           *d;
    struct dirent *dir;
    //check if the path provided has a slash
    if( folder_name[strlen(folder_name) - 1] != '/') {
        strcat(folder_name, "/");
    }
    d = opendir(folder_name);
    if (d) {
        while ((dir = readdir(d)) != NULL)
        {
            char file_path[strlen(folder_name) +  strlen(dir->d_name) + 2];
            if( strcmp(dir->d_name, ".") == 0 || strcmp(dir->d_name, "..") == 0 ) {
                continue;
            }
            //build the path
            strcpy( file_path, folder_name);
            strcat( file_path, dir->d_name);
            //check if file is a folder
            if( opendir(file_path) ){
                //call recursively this function
                get_files(file_path);
            } else {
                insert_into_array(file_path);
            }

        }

        closedir(d);
    }

}

static void insert_into_array(char *str){
    char *node_str = (char *) malloc(strlen(str) +1);
    strcpy(node_str, str);
    node* new = malloc(sizeof(node));
    new->path = node_str;
    new->next = NULL;
    if( head == NULL){
        head = new;
        cursor = head;
    } else{
        cursor->next = new;
        cursor = new;
    }
}
